main/distance_sensor.cpp

```diff
@@ -0,0 +1,160 @@
+    }
+}
+
+std::array<meters, DistanceSensors::sensor_count> DistanceSensors::read_all()
+{
+    static std::array<meters, sensor_count> measurements;
+    for (std::size_t i = 0; i < sensor_count; i++)
+    {
+        m_mux.setPort(m_sensors[i].port);
+        while (!m_current_distance_sensor.checkForDataReady())
+        {
+            delay(1);
+        }
+
+        // Get the result of the measurement from the sensor:
+        m_sensors[i].distance = m_current_distance_sensor.getDistance();
+        m_current_distance_sensor.clearInterrupt();
+        measurements[i] = unit_cast<meters>(millimeters(m_sensors[i].distance));
+    }
+
+    return measurements;
+}
+
+static auto predict_distance(const Position &pos, std::size_t sensor_index, const std::span<const Segment> &maze_map)
+    noexcept
+{
+    const auto ray_x = pos.x.get() + sensor_disposition[sensor_index].first.get() * std::cos(pos.theta)
+                     - sensor_disposition[sensor_index].second.get() * std::sin(pos.theta);
+    const auto ray_y = pos.y.get() + sensor_disposition[sensor_index].first.get() * std::sin(pos.theta)
+                     + sensor_disposition[sensor_index].second.get() * std::cos(pos.theta);
+    const auto sensor_angle = pos.theta + sensor_angles[sensor_index];
+    const Segment sensor_ray{
+        Eigen::Vector2f{ray_x.count(), ray_y.count()},
+        Eigen::Vector2f{
+            (ray_x + max_predict_range * std::cos(sensor_angle)).count(),
+            (ray_y + max_predict_range * std::sin(sensor_angle)).count(),
+        },
+    };
+
+    auto make_result = [&](const auto &wall)
+    { return std::pair<float, const Segment &>(sensor_ray.intersection_distance(wall), wall); };
+
+    auto all_distances = maze_map | std::views::transform(make_result);
+    return *std::ranges::min_element(all_distances, [](const auto &a, const auto &b) { return a.first < b.first; });
+}
+
+std::pair<DistanceSensors::Measurements, DistanceSensors::Jacobian> DistanceSensors::predict(
+    const Position &pos,
+    const std::span<const Segment> &maze_map
+) noexcept
+{
+    Measurements error = Measurements::Zero();
+    Jacobian jacobian = Jacobian::Zero();
+    for (std::size_t i = 0; i < sensor_count; i++)
+    {
+        const auto measured = unit_cast<meters>(millimeters(m_sensors[i].distance)).count();
+        if (measured > max_sensor_range.count())
+        {
+            continue;
+        }
+
+        const auto [distance, wall] = predict_distance(pos, i, maze_map);
+        if (distance > max_predict_range.count())
+        {
+            continue;
+        }
+
+        const auto wall_coefficients = wall.line().coeffs();
+        const auto A = wall_coefficients.x();
+        const auto B = wall_coefficients.y();
+        const auto C = wall_coefficients.z();
+        const auto theta = pos.theta + sensor_angles[i];
+        const auto denominator = A * std::cos(theta) + B * std::sin(theta);
+
+        error(i) = measured - distance;
```

Comment:
I think it's better to keep the values with units up to here.

```suggestion
        error(i) = (measured - distance).count();
```